docs/upgrade.rst

```diff
@@ -8,6 +8,20 @@ to a new version. But it goes into more detail. This includes things like
 deprecated APIs and their replacements, build system changes, general code
 modernization and other useful information.
 
+.. _upgrade-guide-2.10:
+
+v2.10
+=====
+
+``py::gil_scoped_acquire`` & ``py::gil_scoped_release`` in pybind11 versions
+< v2.10.1 do not support nested access. In v2.10.1, a configuration option
+``PYBIND11_SIMPLE_GIL_MANAGEMENT`` was added, defaulting to ``OFF``. The
+simpler implementations support nested access, but do not support dissociation
```

Comment:
```suggestion
simpler implementations support nested access, but do not support dissociation
```

I'm a little worried about promising "nested access is supported" when we have exactly 0 tests for it. Possibly "should support nested access" for now?

Reply:
There is a **huge** difference: the simple version is just a pair of calls to `PyEval_SaveThread()` and `PyEval_RestoreThread()`. Those are exhaustively exercised where they are implemented, in core Python.

What you are asking for is similar to "`Py_INCREF()` should be supported."

The non-simple version is a very complicated implementation that does not have corresponding unit tests in the project that introduces it (i.e. in pybind11).

Reply:
That still doesn't mean we know it works in pybind11 - we didn't add tests. I've been at this too long to believe we can promise untested features. There could very well be some weird behavior on some weird compiler that is pybind11 specific. I'd prefer the word "should" be included if we don't test - we can't promise without tests. If we didn't test `Py_INCREF()`, I'd put "should" on that too (though that would be on the pybind11 wrapping, not the upstream feature - that would be more like adding a `py::incref()` and not testing it and promising it works).

I'd say it probably works, but we did not earn the right to say it works with certainty - that requires a pybind11 test.

Reply:
(I also agree it's more likely to work than dissociation - but to be fair, there are at least two major projects with thousands of users each that use this feature _in_ pybind11, while there are no nested access pybind11 users at all yet - the feature wasn't supported / didn't work before.)

Reply:
PS: this is also a nit that you have permission to ignore - I think it should be "should", but it's not going to make or break anything. :)

Reply:
To be clear, by "nesting" we mean:

```
gil_scoped_release release;
gil_scoped_acquire acquire;
// do something with GIL here
```

Certainly worth adding a test for that.

TensorStore has been using its own gil acquire/release classes that are equivalent to the "simple" version, and does rely on nesting as defined above for its pickling operations, among other things:
https://github.com/google/tensorstore/blob/master/python/tensorstore/serialization.cc

Reply:
Thanks @jbms for making that so concrete! I'll give that a try, looks simple enough, then we can keep the sentence.

Otherwise I agree the sentence is an invalid logical leap from "nesting does not work with B" to "nesting does work with A", we could rewrite it to the former, but adding the test is better.

Reply:
Without using this change, the above would not work at all, or just be flaky (probably when threading)?

I'd much rather have a test than runs in pybind11's CI for something we promise, but okay, downstream works too (and is no worse what what we used for the dissociate feature in the past, apparently).

Reply:
Sorry, my comment above about nesting was supposed to be more of a question rather than a statement, but I left off a question mark.

That is the simplest form of nesting, and that should work both with the "non-simple" and "simple" forms of gil_scoped_{acquire,release}.

The real test is nesting where we combine both the pybind11 apis and the CPython PyGILState_Ensure, as in
https://github.com/pybind/pybind11/issues/1276#issuecomment-366261431

Reply:
According to @wjakob in that thread, the custom implementation is supposed to be compatible (at least it was as of 2.7? Maybe not now). The issue that prompted this is https://github.com/pytorch/pytorch/issues/83101 - that's why we started on this "simple version" in the first place. There it looks exactly like what you have above, except it is running a bunch of Python to import PyTorch inside the "do stuff here" part.

Reply:
@henryiii I'm thinking we could unblock the release by completely omitting this section here. I think this PR is otherwise good to go as-is: it's nice to have the extra option, but because we don't have a reproducer (in the CI) for the reported problem, it would be premature to say anything about the future. For this release, we could just briefly mention this PR in the changelog, like any other PR.

In summary: Release first, continue work on a reproducer, make new plans based on the outcome.

Reply:
I think that's a good idea (and why I didn't have a timeline in the original version). This is still a nice option, having it in a released version will make it much easier for downstream projects to test it out, and even if our complex version is perfect or fixable, I rather like that the simple version is also already there for PyPy. We can announce upgrade plans in 2.11.0 if we go with it.